### ISSUE TITLE
Wrong formula for CosineAnnealingLR

### ISSUE THREAD

**Issue by bbbbbbbbba** [pytorch/pytorch#152081]
### 📚 The doc issue

https://github.com/pytorch/pytorch/blob/1eba9b3aa3c43f86f4a2c807ac8e12c4a7767340/torch/optim/lr_scheduler.py#L1054-L1056

This formula does not incorporate the learning rate of the last step, is the same as the "If the learning rate is set solely by this scheduler" formula below, and does not seem to agree with the actual calculation for this case:

https://github.com/pytorch/pytorch/blob/1eba9b3aa3c43f86f4a2c807ac8e12c4a7767340/torch/optim/lr_scheduler.py#L1123-L1129

### Suggest a potential alternative/fix

I think the correct formula should be something like:
```
            \eta_{t+1} & = \eta_{min} + (\eta_t - \eta_{min})\left.\left(1
            + \cos\left(\frac{T_{cur}+1}{T_{max}}\pi\right)\right)\middle/\left(1
            + \cos\left(\frac{T_{cur}}{T_{max}}\pi\right)\right)\right.,
            & T_{cur} \neq (2k+1)T_{max}; \\
```

**Comment by janeyx99** [comment]
I'd review a PR fixing this, though I haven't looked closely and would am curious difference between `t` and `cur`. Are they the same?

**Comment by bbbbbbbbba** [comment]
To be honest I'm not completely clear either, but I'll guess that maybe `T_{cur}` is the "absolute" time which refers to "the number of epochs since the last restart in SGDR" (according to the documentation), whereas `t` is a "relative" time in that you can set any moment as `t=0`. For example, if you are continuing another training session, you may want to call the first epoch of the current session `t=0`, but `T_cur=0` should refer to the overall first epoch when the weights are randomly initialized.

I'm not really an expert in NN training, so it would be best if an expert could confirm!

**Comment by Jacobgoss30** [comment]
Hi everyone — I’ve opened a PR to update the CosineAnnealingLR docstring to better reflect how the scheduler actually behaves: https://github.com/pytorch/pytorch/pull/152936.

The current docstring describes the learning rate using the SGDR closed-form cosine schedule, but the implementation in get_lr() uses a recursive formula to compute the learning rate at each step. This mismatch can lead to confusion, especially for users expecting the learning rate to follow the theoretical SGDR pattern exactly.

In addition to updating the docstring to show the correct recursive form, I also clarified the meaning of the variable T_cur. In the theory this is the number of epochs since last restart. However, since this implementation doesn’t support restarts, T_cur increases monotonically and is effectively the same as t, the current epoch index.

The updated docstring now:

Starts with the recursive formula (which is what the scheduler actually uses)

Retains the SGDR closed-form expression for reference

Clearly explains the relationship between t and T_cur

Removes misleading references to warm restarts, which are not implemented

Let me know what you think

**Comment by bbbbbbbbba** [comment]
Shouldn't we mention that `T_{cur} = (2k+1)T_{max}` is still a special case? (Otherwise it would be division by zero.)
https://github.com/pytorch/pytorch/blob/1eba9b3aa3c43f86f4a2c807ac8e12c4a7767340/torch/optim/lr_scheduler.py#L1117-L1122

**Comment by janeyx99** [comment]
@bbbbbbbbba I'd review a docs PR clarifying that
